tests/unit/test_tls_lib.py

```diff
@@ -0,0 +1,311 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import unittest
+from unittest import mock
+from unittest.mock import patch
+
+from ops.testing import Harness
+
+from charm import MongodbOperatorCharm
+from tests.unit.helpers import patch_network_get
+
+RELATION_NAME = "certificates"
+
+
+class TestMongoTLS(unittest.TestCase):
+    @patch_network_get(private_address="1.1.1.1")
+    def setUp(self):
+        self.harness = Harness(MongodbOperatorCharm)
+        self.harness.begin()
+        self.harness.add_relation("database-peers", "mongodb-peers")
+        self.harness.set_leader(True)
+        self.charm = self.harness.charm
+        self.addCleanup(self.harness.cleanup)
+
+    @patch_network_get(private_address="1.1.1.1")
+    def test_on_set_tls_private_key(self):
+        """Tests setting of TLS private key in external/internal certificate scenarios.
+
+        Note: this implicitly tests: _request_certificate & _parse_tls_file
+        """
+        # Tests for leader unit (ie internal certificates and external certificates)
+        self.harness.set_leader(True)
+        action_event = mock.Mock()
+        action_event.params = {}
+
+        # generated rsa key test - leader
+        self.harness.charm.tls._on_set_tls_private_key(action_event)
+        self.verify_internal_rsa_csr()
+        self.verify_external_rsa_csr()
+
+        # provided rsa key test - leader
+        set_app_rsa_key = self.harness.charm.app_peer_data["key"]
+        action_event.params = {"internal-key": set_app_rsa_key}
```

Comment:
Is this expected behavior? Where we can `set_tls_private_key` with the same key as it already had, and that works?

I haven't read the other code though, is `key` something different?

Reply:
I think I'm just not clear what's being tested here, as we're getting `key` from relation data, updating key, then checking that the `internal_rsa_csr` is there, which it is because it was there already? I'm sure I'm missing something though.

Reply:
Hi Marc great questions here, I only used `set_app_rsa_key = self.harness.charm.app_peer_data["key"]` as it was in the correct format that is needed by the `set-key` action so this test was just meant to test that setting the key to a specified value works, but I get what you're saying about it being the same I can change it. `set-key` for leaders will reset both the `internal` and `external` keys so if you don't pass in a value it gets generated randomly.